core/sile.lua

```diff
@@ -240,10 +241,251 @@ function SILE.resolveFile(filename, pathprefix)
   return nil
 end
 
+SILE.currentCommandStack = {
+  --- Internal: push-pop balance checking ID
+  _lastPushId = 0,
+  --- Internal: Function assigned to frames to convert them to string
+  _frameToString = function(self, skipFile, withAttrs)
+    local str
+    if skipFile or not self.file then
+      str = ""
+    else
+      str = self.file .. " "
+    end
+
+    if self.line then
+      str = str .. "l." .. self.line .. " "
+      if self.column then
+        str = str .. "col." .. self.column .. " "
+      end
+    end
+
+    if self.tag then
+      -- Command
+      str = str .. "\\" .. self.tag
+      if withAttrs then
+        str = str .. "["
+        local first = true
+        for key, value in pairs(self.options) do
+          if first then first = false else str = str .. ", " end
+          str = str .. key .. "=" .. value
+        end
+        str = str .. "]"
+      end
+    elseif self.text then
+      -- Literal string
+      local text = self.text
+      if text:len() > 20 then
+        text = text:sub(1, 18) .. "…"
+      end
+      text = text:gsub("\n", "␤"):gsub("\t", "␉"):gsub("\v", "␋")
+
+      str = str .. "\"" .. text .. "\""
+    else
+      -- Unknown
+      str = str .. type(self.content) .. "(" .. self.content .. ")"
+    end
+
+    --str = str .. "\n\n" .. self.pushTrace .. "\n"
+
+    return str
+  end
+}
+--- Push a command frame on to the stack to record the execution trace for debugging.
+--- Carries information about the command call, not the command itself.
+--- Must be popped with `pop(returnOfPush)`.
+function SILE.currentCommandStack:pushCommand(tag, line, column, options, file)
+  return self:_pushFrame({
+    tag = tag or "???",
+    file = file or SILE.currentlyProcessingFile,
+    line = line,
+    column = column,
+    options = options or {}
+  })
+end
+--- Push a command frame on to the stack to record the execution trace for debugging.
+--- Command arguments are inferred from AST content, any item may be overridden.
+--- Must be popped with `pop(returnOfPush)`.
+function SILE.currentCommandStack:pushContent(content, tag, line, column, options, file)
+  return self:_pushFrame({
+    tag = tag or (type(content) == "table" and content.tag) or "???",
```

Comment:
```suggestion
    tag = tag or content.tag or "???",
```
I don't think you have to do all the type testing here, if content isn't a table `.tag` won't exist an evaluate to `nil` anyway, allowing you to keep moving down the flow control chain. 

Also should this particular instance throw a `SU.warn()`? What scenario could through the third option here?

Reply:
I do believe it is needed, though only in edge cases:
```lua
> content=false
> content or content.tag
stdin:1: attempt to index a boolean value (global 'content')
stack traceback:
	stdin:1: in main chunk
	[C]: in ?
```

Warning seems like a sensible idea.

Reply:
I understand the potential Lua edge case but I don't see such an edge case in our usage. Looking at the code that _calls_ this command it is **already assuming** `content` is a table. If it's not a table this code shouldn't get called at all. If we are supposed to be passed a content table (even an empty one) then we should assume we can treat it accordingly. If we get the wrong data type we should raise an error about it and die anyway rather than silently suppressing an error condition.

Reply:
The philosophy of this code is, that it is non-essential to the typesetting process and that it should never cause a crash. I can't predict how will the code evolve and I'd rather print a nonsencical trace than crash an otherwise valid program. This would be a lot easier to enforce in a statically typed language, but alas.

Reply:
I understand that goal, but I think in some ways your code (was) working against that. Unfortunately with Lua not being statically typed you can't enforce good practice or easily catch mistakes at compile time. I can understand not wanting to crash inside the tracer, but I think unnecessary code complexity is almost a worse enemy. I'd rather start causing a crash early in the development process if we change something and start passing around the the wrong type of content than silently hide the fact that the content type changed and potentially be messed up for a long time before anybody catches the miss-match.

If you really want to be robust, test to make sure it is a table right out of the gate and `SU.error()` if not. Then after that you can code assuming it is a table. In this case I don't even really recommend that because we don't have any way to test the error condition because we would be crashing elsewhere in the code before even getting to that point, so confirming we coded the error condition would correctly would be tricky.